frege/compiler/gen/java/MethodCall.fr

```diff
@@ -300,39 +307,85 @@ wildReturn g (symv@SymV{})  = [ v | v@TVar{} ← values (U.freeTauTVars [] TreeM
-      -- retlazy = if not (null fldstri) then (last fldstri).isStrict else False
-      callit jex = JInvoke { args = [], jex = JExMem { jex, name = "call", targs = [] } }
-      call = if retstrict then id else callit
+--- Generates a method that overrides an interface method and wraps a call to an IR method.
+---
+--- Parameters:
+--- - global
+--- - The expression of the instance of the IR (i.e. the receiver).
+--- - The Java type of the IR.
+--- - The symbol of the invoked IR method.
+--- - The name of the invoked IR method.
+--- - The symbol information on the native member function.
+--- - The "java item" of the native member function.
+-- TODO make sure that the every field of an IR is 1-or-more arity functions beforehand
+wrapIRMethod :: Global -> JExpr -> JType -> Symbol -> String -> SymInfo8 -> String -> JDecl
+wrapIRMethod g this irjt fldsym fldnm nativsi nativnm =
+  let -- collect the FormalArgs of the native member function, excluding the receiver
+      nativargs = argDefs attrFinal (nativsi.{ argSigs <- tail, argJTs <- tail }) (getArgs g)
+      -- the tau of the result type of the IR field
+      fldrettau = fst $ U.returnType fldsym.typ.rho
+      -- the return type of the overridden method
+      methretjt = case unST fldrettau of
+        Just (_, a)
+          | isJust $ U.isUnit a -> nativ "void" []
+          | otherwise           -> tauJT g a
+        Nothing -> nativsi.returnJT
+      -- force an expression. if it is an ST, run it after that
+      call = (if isST then tstRun (boxedv methretjt) else id) . callit where
+        isST = isJust $ unST fldrettau
+        -- add ".call()" after jex
+        callit jex = JInvoke { args = [], jex = JExMem { jex, name = "call", targs = [] } }
+        -- box JType, but box void into Short
+        boxedv (v@Nativ{typ="void"}) = boxed (v.{typ = "short"})
+        boxedv jt = boxed jt
+      -- "return" if the return type is not the primitive void
+      ret = if isVoid methretjt then JEx else JReturn where
+        isVoid Nativ{typ="void"} = true
+        isVoid _                 = false
   in
   JMethod
   { attr = attrs [JPublic]
   , gvars = []
-  , jtype = nativsi.returnJT
+  , jtype = methretjt
   , name = nativnm
   , args = nativargs
-  , body = JBlock $ pure $ JReturn $ call $ invokeIR irjt fldnm this $
-      zipWith (\(_, _, _, x) s -> StriArg x s.isStrict) nativargs fldstri
+  , body = JBlock $ pure $ ret $ call $ invokeIR irjt fldnm this $
+      map (\(_, _, _, x) -> x) nativargs
   }
 
-data StriArg = StriArg { name :: String, strict :: Bool }
-
-striArgExpr :: StriArg -> JExpr
-striArgExpr (StriArg{name, strict}) =
-  let thunkjt = Ref { jname = JName { qual = "", base = "Thunk" }, gargs = [] }  -- TODO use the proper JType
-      lazy s = JInvoke { args = [JAtom s], jex = JStMem { jt = thunkjt, name = "lazy", targs = [] } }
+--- Wraps an expression into Thunk.lazy, without generics arguments:
+--- > Thunk.lazy(expr)
+asLazy :: JExpr -> JExpr
+asLazy expr =
```

Comment:
This is slightly different from `frege.compiler.gen.java.Common.lazyJX`. This function doesn't take a `JType`. A java compiler can infer the generics of `Thunk.lazy` from its argument, so there isn't a need to emit one, or is it?

Reply:
That's probably so, however, I adopted the policy to not rely on Java type inference if at all possible, as the frege compiler should know the desired types anyway.
Or, to put it differently, if we don't know the types, then we might need to re-consider our code generation model.

Therefore, I prefer having everything explicit in general.

That being said, it is possible that we can avoid some clutter here and there in code generation primitives like "lazyJX". But then we should adopt this in general, not just for *some code*.